sql/preprocess/preprocess.go

```diff
@@ -0,0 +1,412 @@
+				p.scanBuf = &t
+				return p.newTextualTok()
+			}
+			return t
+		case ',':
+			t := token{Type: commaTok, Text: ","}
+			if inText {
+				p.scanBuf = &t
+				return p.newTextualTok()
+			}
+			return t
+		case '\t':
+			return token{Type: whitespaceTok, Text: "\t"}
+		case ' ':
+			t := token{Type: whitespaceTok, Text: " "}
+			if inText {
+				p.scanBuf = &t
+				return p.newTextualTok()
+			}
+			return t
+
+		case eof:
+			t := token{Type: eofTok}
+			if inText {
+				p.scanBuf = &t
+				return p.newTextualTok()
+			}
+			return t
+
+		default:
+			p.text.WriteRune(ch)
+			inText = true
+		}
+	}
+}
+
+func (p *processor) unscan(t token) {
+	p.parseBuf = &t
+}
+
+// read reads the next rune from the buffered reader.
+// Returns the rune(0) if an error occurs (or io.EOF is returned).
+func (p *processor) read() rune {
+	ch, _, err := p.r.ReadRune()
+	if err != nil {
+		return eof
+	}
+	return ch
+}
+
+// newTok creates a new token from current processor state
+func (p *processor) newTextualTok() token {
+	// p.buf = buf
+	t := token{
+		Type: textTok,
+		Text: strings.TrimSpace(p.text.String()),
+		Pos:  position{Line: p.line, Col: p.col},
+	}
+
+	// identify keywords
+	switch strings.ToLower(t.Text) {
+	case asTok.String():
+		t.Type = asTok
+	case byTok.String():
+		t.Type = byTok
+	case fromTok.String():
+		t.Type = fromTok
+	case groupTok.String():
+		t.Type = groupTok
+	case havingTok.String():
+		t.Type = havingTok
+	case unionTok.String():
+		t.Type = unionTok
+	case limitTok.String():
+		t.Type = limitTok
+	case orderTok.String():
+		t.Type = orderTok
+	case whereTok.String():
+		t.Type = whereTok
+	}
+
+	return t
+}
+
+// eof represents a marker rune for the end of the reader.
+var eof = rune(0)
+
+// token is a recognized token from the outlineline lexicon
+type token struct {
+	Type tokenType
+	Pos  position
+	Text string
+}
+
+type position struct {
+	Line, Col int
+}
+
+// String implements the stringer interface for token
+func (t token) String() string {
+	return t.Text
+}
+
+// tokenType enumerates the different types of tokens
+type tokenType int
+
+const (
+	// IllegalTok is the default for unrecognized tokens
+	IllegalTok tokenType = iota
+	eofTok
+
+	literalBegin  // marks the beginning of literal tokens in the token enumeration
+	indentTok     // tab character "\t" or two consecutive spaces"  "
+	newlineTok    // line break
+	textTok       // a token for arbitrary text
+	commaTok      // a "," character
+	leftParenTok  // "("
+	rightParenTok // ")"
+	whitespaceTok // spaces, tabs, newlines
+	quoteTok      // a single quote: '
+	literalEnd    // marks the end of literal tokens in the token enumeration
+
+	keywordBegin // keywordBegin marks the start of SQL keyword tokens in the token enumeration
+	asTok
+	byTok
+	fromTok
+	groupTok
+	havingTok
+	limitTok
+	orderTok
+	unionTok
+	whereTok
+	keywordEnd // keywordEnd marks the end of keyword tokens in the token enumeration
+)
+
+func (t tokenType) String() string {
```

Comment:
I think it would be a good idea to have a complete list of "reserved words" here and then just noop or throw on them when parsing, otherwise things like "join" could get caught as a table name instead of an operator. And might be better for us to throw here an "unsuported operation error" than having it explode on the user from deeper in the stack.